File: src/xianyu_cli/core/api_client.py
"""Core HTTP API client for the Goofish/XianYu mtop gateway.

All API calls go through: https://h5api.m.goofish.com/h5/{api_name}/{version}/
Sign: MD5(token + "&" + timestamp + "&" + appKey + "&" + data)
"""

from __future__ import annotations

import json
import logging
from typing import Any

import httpx

from xianyu_cli.core.sign import extract_token, generate_sign, get_timestamp
from xianyu_cli.utils._common import API_BASE_URL, APP_KEY
from xianyu_cli.utils.anti_detect import AntiDetect

logger = logging.getLogger(__name__)


class ApiError(Exception):
    """Raised when the mtop API returns an error."""

    def __init__(self, message: str, ret: list[str] | None = None):
        super().__init__(message)
        self.ret = ret or []


class TokenExpiredError(ApiError):
    """Raised when the session token has expired."""


class RateLimitError(ApiError):
    """Raised when rate-limited by the server."""


class GoofishApiClient:
    """HTTP client for the Goofish mtop API gateway.

    Usage:
        async with GoofishApiClient(cookies) as client:
            result = await client.call("mtop.taobao.idlesearch.searchmain", data)
    """

    def __init__(
        self,
        cookies: dict[str, str],
        anti_detect: AntiDetect | None = None,
        timeout: float = 20.0,
        max_retries: int = 3,
    ):
        self.cookies = dict(cookies)
        self.anti_detect = anti_detect or AntiDetect()
        self.timeout = timeout
        self.max_retries = max_retries
        self._client: httpx.AsyncClient | None = None

    async def __aenter__(self) -> GoofishApiClient:
        headers = self.anti_detect.get_headers()
        self._client = httpx.AsyncClient(
            timeout=self.timeout,
            headers=headers,
            follow_redirects=True,
        )
        return self

    async def __aexit__(self, *args: Any) -> None:
        if self._client:
            await self._client.aclose()
            self._client = None

    @property
    def token(self) -> str:
        """Extract token from _m_h5_tk cookie."""
        m_h5_tk = self.cookies.get("_m_h5_tk", "")
        return extract_token(m_h5_tk)

    def _build_cookie_header(self) -> str:
        """Build Cookie header string from cookies dict."""
        return "; ".join(f"{k}={v}" for k, v in self.cookies.items())

    async def call(
        self,
        api: str,
        data: dict[str, Any] | None = None,
        version: str = "1.0",
    ) -> dict[str, Any]:
        """Make a signed API call to the mtop gateway.

        Args:
            api: API name, e.g. "mtop.taobao.idlesearch.searchmain"
            data: Request data dict (will be JSON-encoded).
            version: API version, default "1.0".

        Returns:
            The 'data' field from the API response.

        Raises:
            ApiError: On API-level errors.
            TokenExpiredError: When session token has expired.
            RateLimitError: When rate-limited.
        """
        assert self._client is not None, "Use 'async with' to create the client"

        data = data or {}
        data_str = json.dumps(data, separators=(",", ":"), ensure_ascii=False)

        token_refreshed = False
        last_error: Exception | None = None
        for attempt in range(self.max_retries):
            if attempt > 0:
                await self.anti_detect.backoff_delay(attempt)

            try:
                result = await self._do_request(api, data_str, version)
                return result
            except RateLimitError as e:
                logger.warning("Rate limited on attempt %d: %s", attempt + 1, e)
                last_error = e
                continue
            except TokenExpiredError:
                if not token_refreshed:
                    logger.info("Token expired, attempting auto-refresh")
                    await self.refresh_token()
                    token_refreshed = True
                    continue  # retry with refreshed token
                raise
            except ApiError:
                raise

        raise last_error or ApiError("Max retries exceeded")

    async def _do_request(
        self,
        api: str,
        data_str: str,
        version: str,
    ) -> dict[str, Any]:
        """Execute a single signed API request."""
        assert self._client is not None

        t = get_timestamp()
        sign = generate_sign(self.token, t, data_str)

        url = f"{API_BASE_URL}/{api}/{version}/"

        params = {
            "jsv": "2.7.2",
            "appKey": APP_KEY,
            "t": t,
            "sign": sign,
            "v": version,
            "type": "originaljson",
            "dataType": "json",
            "timeout": "20000",
            "api": api,
            "sessionOption": "AutoLoginOnly",
        }

        headers = {
            "Cookie": self._build_cookie_header(),
        }

        # Anti-detection jitter
        await self.anti_detect.jitter_delay()

        resp = await self._client.post(
            url,
            params=params,
            data={"data": data_str},
            headers=headers,
        )

        # Update cookies from response Set-Cookie headers
        self._update_cookies_from_set_cookie(resp)

        resp.raise_for_status()
        body = resp.json()

        return self._parse_response(body)

    def _update_cookies_from_response(self, resp: httpx.Response) -> None:
        """Capture and store any Set-Cookie headers from the response."""
        for key, value in resp.cookies.items():
            self.cookies[key] = value

    def _parse_response(self, body: dict[str, Any]) -> dict[str, Any]:
        """Parse the mtop API response, checking for errors."""
        ret = body.get("ret", [])
        ret_str = " ".join(ret) if isinstance(ret, list) else str(ret)

        # Check for success
        if any("SUCCESS" in r for r in ret):
            return body.get("data", {})

        # Check for token expiry
        if any("TOKEN_EXOIRED" in r or "TOKEN_EXPIRED" in r or "FAIL_SYS_TOKEN" in r for r in ret):
            raise TokenExpiredError(f"Token expired: {ret_str}", ret)

        # Check for rate limiting
        if any("FAIL_SYS_ILLEGAL_ACCESS" in r or "RGV587" in r for r in ret):
            raise RateLimitError(f"Rate limited: {ret_str}", ret)

        raise ApiError(f"API error: {ret_str}", ret)

    def _update_cookies_from_set_cookie(self, resp: httpx.Response) -> None:
        """Parse Set-Cookie headers directly from the response.

        When cookies are sent via the Cookie header (bypassing httpx's jar),
        resp.cookies may not capture Set-Cookie values. Parse raw headers.
        """
        self._update_cookies_from_response(resp)
        for raw in resp.headers.get_list("set-cookie"):
            if "=" in raw:
                name_value = raw.split(";")[0]
                name, _, value = name_value.partition("=")
                name = name.strip()
                value = value.strip()
                if name and value:
                    self.cookies[name] = value

    async def refresh_token(self) -> None:
        """Refresh m_h5_tk by calling the index API.

        Uses a fresh HTTP client to avoid cookie jar interference, and
        includes the minimum mtop query params the server expects.
        """
        url = (
            f"{API_BASE_URL}/"
            "mtop.gaia.nodejs.gaia.idle.data.gw.v2.index.get/1.0/"
        )

        params = {
            "jsv": "2.7.2",
            "appKey": APP_KEY,
            "type": "originaljson",
            "dataType": "json",
        }

        headers = {
            "Cookie": self._build_cookie_header(),
            "Referer": "https://www.goofish.com/",
        }

        # Use a fresh client to avoid cookie jar interference
        async with httpx.AsyncClient(
            follow_redirects=True,
            timeout=15.0,
            headers=self.anti_detect.get_headers(),
        ) as fresh:
            # Try h5api first, then main site, then h5api again
            for attempt_url in [url, "https://www.goofish.com/", url]:
                p = params if "h5api" in attempt_url else None
                try:
                    resp = await fresh.get(
                        attempt_url, headers=headers, params=p
                    )
                    self._update_cookies_from_set_cookie(resp)
                    headers["Cookie"] = self._build_cookie_header()
                    if "_m_h5_tk" in self.cookies:
                        break
                except Exception:
                    logger.debug("Refresh attempt failed: %s", attempt_url)

        logger.info("Token refreshed, _m_h5_tk present: %s", "_m_h5_tk" in self.cookies)

File: golang/core/api_client.go
package core

import (
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"strings"

	"xianyu-cli/utils"
)

// API error types
type ApiError struct {
	Message string
	Ret     []string
}

func (e *ApiError) Error() string { return e.Message }

type TokenExpiredError struct{ ApiError }
type RateLimitError struct{ ApiError }

// GoofishApiClient is the HTTP client for the Goofish mtop API gateway.
type GoofishApiClient struct {
	Cookies        map[string]string
	AntiDetect     *utils.AntiDetect
	Timeout        int
	MaxRetries     int
	TokenRefreshed bool
	client         *http.Client
}

// NewGoofishApiClient creates a new API client.
func NewGoofishApiClient(cookies map[string]string) *GoofishApiClient {
	// Copy cookies to avoid mutation
	cookiesCopy := make(map[string]string)
	for k, v := range cookies {
		cookiesCopy[k] = v
	}

	transport := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	}
	if utils.ProxyURL != "" {
		proxyURL, err := url.Parse(utils.ProxyURL)
		if err == nil {
			transport.Proxy = http.ProxyURL(proxyURL)
		}
	}

	return &GoofishApiClient{
		Cookies:    cookiesCopy,
		AntiDetect: utils.NewAntiDetect(),
		Timeout:    20,
		MaxRetries: 3,
		client: &http.Client{
			Transport: transport,
			CheckRedirect: func(req *http.Request, via []*http.Request) error {
				return nil // follow redirects
			},
		},
	}
}

// Token extracts the token from _m_h5_tk cookie.
func (c *GoofishApiClient) Token() string {
	return ExtractToken(c.Cookies["_m_h5_tk"])
}

// Call makes a signed API call to the mtop gateway.
func (c *GoofishApiClient) Call(api string, data map[string]interface{}, version string) (map[string]interface{}, error) {
	if version == "" {
		version = "1.0"
	}
	if data == nil {
		data = make(map[string]interface{})
	}

	dataBytes, _ := json.Marshal(data)
	dataStr := string(dataBytes)

	tokenRefreshed := false
	var lastErr error

	for attempt := 0; attempt < c.MaxRetries; attempt++ {
		if attempt > 0 {
			c.AntiDetect.BackoffDelay(attempt)
		}

		result, err := c.doRequest(api, dataStr, version)
		if err == nil {
			return result, nil
		}

		switch e := err.(type) {
		case *RateLimitError:
			log.Printf("Rate limited on attempt %d: %s", attempt+1, e.Message)
			lastErr = e
			continue
		case *TokenExpiredError:
			if !tokenRefreshed {
				log.Printf("Token expired, attempting auto-refresh")
				c.RefreshToken()
				tokenRefreshed = true
				continue
			}
			return nil, err
		case *ApiError:
			return nil, err
		default:
			return nil, err
		}
	}

	if lastErr != nil {
		return nil, lastErr
	}
	return nil, &ApiError{Message: "Max retries exceeded"}
}

func (c *GoofishApiClient) doRequest(api, dataStr, version string) (map[string]interface{}, error) {
	t := GetTimestamp()
	sign := GenerateSign(c.Token(), t, dataStr)

	apiURL := fmt.Sprintf("%s/%s/%s/", utils.APIBaseURL, api, version)

	params := url.Values{}
	params.Set("jsv", "2.7.2")
	params.Set("appKey", utils.AppKey)
	params.Set("t", t)
	params.Set("sign", sign)
	params.Set("v", version)
	params.Set("type", "originaljson")
	params.Set("dataType", "json")
	params.Set("timeout", "20000")
	params.Set("api", api)
	params.Set("sessionOption", "AutoLoginOnly")

	fullURL := apiURL + "?" + params.Encode()

	// Anti-detection jitter
	c.AntiDetect.JitterDelay()

	body := url.Values{}
	body.Set("data", dataStr)

	req, err := http.NewRequest("POST", fullURL, strings.NewReader(body.Encode()))
	if err != nil {
		return nil, err
	}

	// Set headers
	for k, vals := range utils.DefaultHeaders() {
		for _, v := range vals {
			req.Header.Set(k, v)
		}
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Cookie", utils.BuildCookieHeader(c.Cookies))

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	// Capture Set-Cookie headers to update cookies (especially unb)
	collectSetCookies(resp, c.Cookies)

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("failed to parse API response: %w", err)
	}

	return c.parseResponse(result)
}

func (c *GoofishApiClient) parseResponse(body map[string]interface{}) (map[string]interface{}, error) {
	retRaw := body["ret"]
	var ret []string
	switch v := retRaw.(type) {
	case []interface{}:
		for _, r := range v {
			ret = append(ret, fmt.Sprint(r))
		}
	case string:
		ret = []string{v}
	}

	retStr := strings.Join(ret, " ")

	// Check for success
	for _, r := range ret {
		if strings.Contains(r, "SUCCESS") {
			if data, ok := body["data"].(map[string]interface{}); ok {
				return data, nil
			}
			return make(map[string]interface{}), nil
		}
	}

	// Check for token expiry
	for _, r := range ret {
		if strings.Contains(r, "TOKEN_EXOIRED") || strings.Contains(r, "TOKEN_EXPIRED") || strings.Contains(r, "FAIL_SYS_TOKEN") {
			return nil, &TokenExpiredError{ApiError{Message: "Token expired: " + retStr, Ret: ret}}
		}
	}

	// Check for rate limiting
	for _, r := range ret {
		if strings.Contains(r, "FAIL_SYS_ILLEGAL_ACCESS") || strings.Contains(r, "RGV587") {
			return nil, &RateLimitError{ApiError{Message: "Rate limited: " + retStr, Ret: ret}}
		}
	}

	return nil, &ApiError{Message: "API error: " + retStr, Ret: ret}
}

// RefreshToken refreshes m_h5_tk by calling the index API.
func (c *GoofishApiClient) RefreshToken() {
	refreshURL := utils.APIBaseURL + "/mtop.gaia.nodejs.gaia.idle.data.gw.v2.index.get/1.0/"

	params := url.Values{}
	params.Set("jsv", "2.7.2")
	params.Set("appKey", utils.AppKey)
	params.Set("type", "originaljson")
	params.Set("dataType", "json")

	urls := []string{
		refreshURL + "?" + params.Encode(),
		"https://www.goofish.com/",
		refreshURL + "?" + params.Encode(),
	}

	transport := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	}
	if utils.ProxyURL != "" {
		proxyURL, _ := url.Parse(utils.ProxyURL)
		transport.Proxy = http.ProxyURL(proxyURL)
	}
	freshClient := &http.Client{Transport: transport}

	cookieHeader := utils.BuildCookieHeader(c.Cookies)

	for _, u := range urls {
		req, err := http.NewRequest("GET", u, nil)
		if err != nil {
			continue
		}
		for k, vals := range utils.DefaultHeaders() {
			for _, v := range vals {
				req.Header.Set(k, v)
			}
		}
		req.Header.Set("Cookie", cookieHeader)
		req.Header.Set("Referer", "https://www.goofish.com/")

		resp, err := freshClient.Do(req)
		if err != nil {
			continue
		}

		// Extract _m_h5_tk from Set-Cookie headers
		for _, cookie := range resp.Cookies() {
			if cookie.Name == "_m_h5_tk" || cookie.Name == "_m_h5_tk_enc" {
				c.Cookies[cookie.Name] = cookie.Value
			}
		}
		// Also parse raw Set-Cookie headers
		for _, raw := range resp.Header["Set-Cookie"] {
			if idx := strings.Index(raw, "="); idx > 0 {
				nameValue := strings.SplitN(raw, ";", 2)[0]
				parts := strings.SplitN(nameValue, "=", 2)
				if len(parts) == 2 {
					name := strings.TrimSpace(parts[0])
					value := strings.TrimSpace(parts[1])
					if (name == "_m_h5_tk" || name == "_m_h5_tk_enc") && value != "" {
						c.Cookies[name] = value
					}
				}
			}
		}
		resp.Body.Close()
		cookieHeader = utils.BuildCookieHeader(c.Cookies)

		if _, ok := c.Cookies["_m_h5_tk"]; ok {
			c.TokenRefreshed = true
			break
		}
	}
}

// Close cleans up the client resources.
func (c *GoofishApiClient) Close() {
	c.client.CloseIdleConnections()
}
